HZC-960: C++ Sample Client (#26)
hazelcast/cloud-docs: docs/modules/ROOT/pages/cpp-client.adoc

= C++ Client
:url-github-cpp: https://github.com/hazelcast/hazelcast-cpp-client/blob/v4.1.0/Reference_Manual.md

== Quick Setup

Follow the below instructions to connect a .Net client to your cluster.

. xref:create-starter-cluster.adoc[Create a cluster].
+
The cluster-specific page will load and cluster's state will be "Pending" for some seconds, as shown below.
+
image:create-cluster-pending.png[A cluster being created]

. When the cluster's state becomes "Running", the previously inactive button `Configure Client` on this page becomes active. Click on it and the following dialog box appears:
+
image:cpp-client-dialog.png[Instructions for installing a .Net client]

. Click on the C++ tab (the Java tab is active by default) and download the ZIP file as instructed in Step 1 of the dialog box.

. Extract the ZIP file and run the command we provided, in the extracted folder, shown in Step 2.

* Windows users may need update this command according to configuration of their system.  

> ** If your target triple doesn't match with operation system, you may need to install Client with providing target suffix. For example: `vcpkg install "hazelcast-cpp-client[openssl]:x64-windows"`

Author: Typo, it should be `target triplet` You can check what is it from here: https://wiki.osdev.org/Target_Triplet

> ** If your target triple doesn't match with operation system, you may need to install Client with providing target suffix. For example: `vcpkg install "hazelcast-cpp-client[openssl]:x64-windows"`

Reviewer: Ah cool, thanks 😄  In that case we can say
```suggestion
** If the link:https://wiki.osdev.org/Target_Triplet[target triplet] doesn't match your operation system, you may need to provide a target suffix to the `install` command. For example: `vcpkg install "hazelcast-cpp-client[openssl]:x64-windows"`
```

> ** Location of the compiled binary may change according to configuration and system defaults.

Author: Yes I tried this on both macOS and Windows and compiled binary was in different direction. It applies only to Windows users.

    auto hazelcastClient = hazelcast::new_client(std::move(config)).get();  // <4>
    auto map = hazelcastClient.get_map("map").get(); // <5>
    auto check = map->put<std::string, std::string>("key", "value").get(); // <6>
    map->clear().get();
    std::cout << "Connection Successful!" << std::endl;
    std::cout << "Now, 'map' will be filled with random entries." << std::endl;

    int iterationCount = 0;
    while (true) { // <7>
        int randomKey = rand();
        std::string randomKeyString = std::to_string(randomKey); 
        try {
            map->put<std::string, std::string>("key" + randomKeyString, "value" + randomKeyString).get();
        } catch (std::exception &e) {
            std::cout << "Put operation failed error:" << e.what() << std::endl;
        }
        if (++iterationCount % 10 == 0){
            std::cout << "Map size:" + std::to_string(map->size().get()) << std::endl;
        }
    }

    return 0;
}
----

<1> First, we set name of the our cluster. 

<2> Then, we enable statistics in order to see statistics in Hazelcast Cloud Console.

<3> Then, we set discovery token of our cluster. 
+

> This discovery token will be used by client in order to discover cluster members.

Author: Done, thanks!

> <4> Now, we create the client with `config`.

Author: Done, thanks!

<5> We get/create a map with name `map`.

> <6> Then, we check add a value to our map, and get it to let the code throw an exception if the value is not correct.

Reviewer: ```suggestion
<6> Then, we add an entry to our map and read it back to check that everything is working.
```

Author: Done, thanks!

    config.get_network_config().get_ssl_config().set_context(std::move(ctx));

    auto hazelcastClient = hazelcast::new_client(std::move(config)).get(); // <5>
    auto map = hazelcastClient.get_map("map").get(); // <6>
    auto check = map->put<std::string, std::string>("key", "value").get(); // <7>
    map->clear().get();
    std::cout << "Connection Successful!" << std::endl;
    std::cout << "Now, 'map' will be filled with random entries." << std::endl;
    
    int iterationCount = 0;
    while (true) { // <8>
        int randomKey = rand();
        std::string randomKeyString = std::to_string(randomKey);
        try {
            map->put<std::string, std::string>("key" + randomKeyString, "value" + randomKeyString).get();
        } catch (std::exception &e) {
            std::cout << "Put operation failed error:" << e.what() << std::endl;
        }
        if (++iterationCount % 10 == 0){
            std::cout << "Map size:" + std::to_string(map->size().get()) << std::endl;
        }
    }

    return 0;
}
----

<1> First, we set name of the our cluster. 

<2> We enable statistics in order to see statistics in Hazelcast Cloud Console.

> <3> We set discovery token of our cluster. 

Reviewer: ```suggestion
<3> Then, we set our cluster's discovery token. 
```

Author: Done, thanks!

+

> This discovery token will be used by client in order to discover cluster members.

Reviewer: ```suggestion
The client will use this discovery token to discover cluster members.
```

Author: Done, thanks!

> <4> We set our certificate file, key file and password in order establish a secure connection with our Cluster..

Reviewer: ```suggestion
<4> To establish a secure connection to our cluster, we set our certificate file, key file and password.
```

Author: Done, thanks!

> <5> We create the client with our `config`.

Reviewer: ```suggestion
<5> We create the client, using our `config` object.
```

Author: Done, thanks!

<6> We get/create a map with name `map`.

> <7> Then, we check add a value to our map, and get it to let the code throw an exception if the value is not correct.

Reviewer: ```suggestion
<7> Then, we add an entry to our map and read it back to check that everything is working.
```

Author: Done, thanks!
